elastic/kibana
# [Saved Objects] Adds config flag to toggle hiddenFromHttpApis SO types conditionally

Fix https://github.com/elastic/kibana/issues/150471
Part of https://github.com/elastic/kibana/issues/149287

This PR adds a config option to set if the Saved Objects HTTP APIs should be accessible to types registered with `hiddenFromHttpApis: true`. 


`allowHttpApiAccess: true` --------> grants access
`allowHttpApiAccess: false` --------> throws in Saved Objects HTTP APIs

`allowHttpApiAccess` can be configured in `kibana.yml` but we don't plan to document the option as it is not intended for public use.

The configuration prevents immediate breaking changes caused by saved objects adopting the option to be [hidden from the HTTP APIs](https://github.com/elastic/kibana/issues/147150) and is a stepping stone toward removing the Saved Objects HTTP APIs.

The option is only applicable to saved objects registered as `hiddenFromHttpApis:true`.



### Checklist
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials. _Code comments should be sufficient to describe the _internal_ behavior._
- [x] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker) _Docker list updated, we don't intend to add user settings in cloud._

### Risk Matrix
| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Code should gracefully handle cases when the HTTP APIs are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service behaving the way intended. |

### For maintainers


### How to test this

Ensure that `allowHttpApiAccess` can be set in `kibana.yml` to override the default:

Add the following debug log line to the SavedObjectsService [setup contract after the `config` is set](https://github.com/elastic/kibana/blob/main/packages/core/saved-objects/core-saved-objects-server-internal/src/saved_objects_service.ts#L125):

<details>
<summary>log config value</summary>

```typescript
public async setup(setupDeps: SavedObjectsSetupDeps): Promise<InternalSavedObjectsServiceSetup> {
    this.logger.debug('Setting up SavedObjects service');

    this.setupDeps = setupDeps;
    const { http, elasticsearch, coreUsageData, deprecations } = setupDeps;

    const savedObjectsConfig = await firstValueFrom(
      this.coreContext.configService.atPath<SavedObjectsConfigType>('savedObjects')
    );
    const savedObjectsMigrationConfig = await firstValueFrom(
      this.coreContext.configService.atPath<SavedObjectsMigrationConfigType>('migrations')
    );
    this.config = new SavedObjectConfig(savedObjectsConfig, savedObjectsMigrationConfig);
   // add the debug log:
    this.logger.debug(
      `SavedObjects default allowHttpApiAccess configuration: ${this.config.allowHttpApiAccess}`
    );
    deprecations.getRegistry('savedObjects').registerDeprecations(
     <...rest>
```

</details>

Configure logging to include the saved objects service logs:
For example:

<details>
<summary>kibana.yml</summary>

```yml
logging:
  appenders:
    so-console:
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "SAVED_OBJECTS_SERVICE--[%date][%level][%logger]---%message"
  loggers:
    - name: savedobjects-service
      level: debug
      appenders: [so-console]
```

</details>

Start kibana (any version of es is fine) and ensure that `allowHttpApiAccess` is set. For example, when run from dev the log should read something like:
`SAVED_OBJECTS_SERVICE--[2023-02-21T12:06:28.561-07:00][DEBUG][savedobjects-service]---SavedObjects default allowHttpApiAccess configuration: false`

In `kibana.yml`, specifically set `savedObjects.allowHttpApiAccess: true` and recheck the logs. Using the same logging config as above, the log line should now read:
`SAVED_OBJECTS_SERVICE--[2023-02-21T12:06:28.561-07:00][DEBUG][savedobjects-service]---SavedObjects default allowHttpApiAccess configuration: true`
